File: app/lib/models/site.dart
import 'dart:math' as math;
import 'package:json_annotation/json_annotation.dart';

part 'site.g.dart';

@JsonSerializable()
class Site {
  final String id;
  final String clientId;
  final String? parentSiteId;
  final String name;
  final String? address;
  final double? centerLatitude;
  final double? centerLongitude;
  final double? boundaryRadius;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool isActive;

  Site({
    required this.id,
    required this.clientId,
    this.parentSiteId,
    required this.name,
    this.address,
    this.centerLatitude,
    this.centerLongitude,
    this.boundaryRadius,
    required this.createdAt,
    required this.updatedAt,
    this.isActive = true,
  });

  factory Site.fromJson(Map<String, dynamic> json) => _$SiteFromJson(json);
  Map<String, dynamic> toJson() => _$SiteToJson(this);

  Site copyWith({
    String? id,
    String? clientId,
    String? parentSiteId,
    String? name,
    String? address,
    double? centerLatitude,
    double? centerLongitude,
    double? boundaryRadius,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isActive,
  }) {
    return Site(
      id: id ?? this.id,
      clientId: clientId ?? this.clientId,
      parentSiteId: parentSiteId ?? this.parentSiteId,
      name: name ?? this.name,
      address: address ?? this.address,
      centerLatitude: centerLatitude ?? this.centerLatitude,
      centerLongitude: centerLongitude ?? this.centerLongitude,
      boundaryRadius: boundaryRadius ?? this.boundaryRadius,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isActive: isActive ?? this.isActive,
    );
  }

  bool get isMainSite => parentSiteId == null;
  bool get isSubSite => parentSiteId != null;

  bool isValidName() {
    return name.isNotEmpty && name.length <= 100;
  }

  bool hasValidCoordinates() {
    if (centerLatitude == null || centerLongitude == null) return true;
    return centerLatitude! >= -90 &&
        centerLatitude! <= 90 &&
        centerLongitude! >= -180 &&
        centerLongitude! <= 180;
  }

  bool containsLocation(double latitude, double longitude) {
    if (centerLatitude == null ||
        centerLongitude == null ||
        boundaryRadius == null) {
      return false;
    }

    final distance = _calculateDistance(
      centerLatitude!,
      centerLongitude!,
      latitude,
      longitude,
    );
    return distance <= boundaryRadius!;
  }

  double _calculateDistance(double lat1, double lon1, double lat2, double lon2) {
    const earthRadius = 6371000.0;
    final dLat = _toRadians(lat2 - lat1);
    final dLon = _toRadians(lon2 - lon1);
    final a = math.sin(dLat / 2) * math.sin(dLat / 2) +
        math.cos(_toRadians(lat1)) *
            math.cos(_toRadians(lat2)) *
            math.sin(dLon / 2) *
            math.sin(dLon / 2);
    final c = 2 * math.asin(math.sqrt(a));
    return earthRadius * c;
  }

  double _toRadians(double degrees) => degrees * (3.14159265359 / 180.0);

  String getHierarchyPath({Site? parentSite}) {
    if (isMainSite) {
      return name;
    }
    if (parentSite != null) {
      return '${parentSite.name} > $name';
    }
    return name;
  }
}

File: api/src/models/site.js
const { DataTypes } = require('sequelize');
const sequelize = require('../database/connection');

const Site = sequelize.define('Site', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  clientId: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'Clients',
      key: 'id',
    },
  },
  parentSiteId: {
    type: DataTypes.UUID,
    allowNull: true,
    references: {
      model: 'Sites',
      key: 'id',
    },
  },
  name: {
    type: DataTypes.STRING(100),
    allowNull: false,
    validate: {
      notEmpty: true,
      len: [1, 100],
    },
  },
  address: {
    type: DataTypes.STRING(500),
    allowNull: true,
  },
  centerLatitude: {
    type: DataTypes.DECIMAL(10, 8),
    allowNull: true,
    validate: {
      min: -90,
      max: 90,
    },
  },
  centerLongitude: {
    type: DataTypes.DECIMAL(11, 8),
    allowNull: true,
    validate: {
      min: -180,
      max: 180,
    },
  },
  boundaryRadius: {
    type: DataTypes.DECIMAL(10, 2),
    allowNull: true,
    validate: {
      min: 0,
    },
  },
  createdAt: {
    type: DataTypes.DATE,
    allowNull: false,
  },
  updatedAt: {
    type: DataTypes.DATE,
    allowNull: false,
  },
  isActive: {
    type: DataTypes.BOOLEAN,
    defaultValue: true,
  },
}, {
  tableName: 'sites',
  timestamps: true,
  indexes: [
    {
      fields: ['clientId', 'name'],
    },
    {
      fields: ['parentSiteId'],
    },
    {
      fields: ['isActive'],
    },
  ],
});

Site.prototype.toJSON = function() {
  const values = { ...this.get() };
  return values;
};

Site.prototype.isMainSite = function() {
  return this.parentSiteId === null;
};

Site.prototype.isSubSite = function() {
  return this.parentSiteId !== null;
};

Site.prototype.containsLocation = function(latitude, longitude) {
  if (!this.centerLatitude || !this.centerLongitude || !this.boundaryRadius) {
    return false;
  }

  const distance = this.calculateDistance(
    this.centerLatitude,
    this.centerLongitude,
    latitude,
    longitude
  );
  return distance <= this.boundaryRadius;
};

Site.prototype.calculateDistance = function(lat1, lon1, lat2, lon2) {
  const earthRadius = 6371000;
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLon = (lon2 - lon1) * Math.PI / 180;
  const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(lat1 * Math.PI / 180) *
    Math.cos(lat2 * Math.PI / 180) *
    Math.sin(dLon / 2) *
    Math.sin(dLon / 2);
  const c = 2 * Math.asin(Math.sqrt(a));
  return earthRadius * c;
};

Site.findByClient = function(clientId, includeInactive = false) {
  const where = { clientId };
  if (!includeInactive) {
    where.isActive = true;
  }
  return this.findAll({
    where,
    order: [['parentSiteId', 'ASC'], ['name', 'ASC']],
  });
};

Site.findMainSites = function(clientId) {
  return this.findAll({
    where: {
      clientId,
      parentSiteId: null,
      isActive: true,
    },
    order: [['name', 'ASC']],
  });
};

Site.findSubSites = function(parentSiteId) {
  return this.findAll({
    where: {
      parentSiteId,
      isActive: true,
    },
    order: [['name', 'ASC']],
  });
};

Site.associate = function(models) {
  Site.belongsTo(models.Client, {
    foreignKey: 'clientId',
    as: 'client',
  });
  Site.belongsTo(models.Site, {
    foreignKey: 'parentSiteId',
    as: 'parentSite',
  });
  Site.hasMany(models.Site, {
    foreignKey: 'parentSiteId',
    as: 'subSites',
  });
  Site.hasMany(models.Equipment, {
    foreignKey: 'siteId',
    as: 'equipment',
  });
};

module.exports = Site;